derive/src/lib.rs

```diff
@@ -70,13 +70,13 @@ fn generate_type(input: TokenStream2) -> Result<TokenStream2> {
     let mut ast: DeriveInput = syn::parse2(input.clone())?;
 
     let ident = &ast.ident;
-    trait_bounds::add(ident, &mut ast.generics, &ast.data)?;
 
     ast.generics
         .lifetimes_mut()
         .for_each(|l| *l = parse_quote!('static));
 
-    let (_, ty_generics, where_clause) = ast.generics.split_for_impl();
+    let (ty_generics, where_clause) = trait_bounds::add(ident, &ast.generics, &ast.data)?;
```

Comment:
This is where it looks a bit odd, we call `add()` and get back more stuff than I'd expect just from reading the method name and arguments. It's fine, better than before even, but just looks a bit odd.
I wonder what this would look like if we just went against orthodoxy and in-lined the whole `trait_bounds::add` function here directly. It would be long, yes, but perhaps more straight forward to read at the end of the day?

Reply:
Yes it does look odd now I have a fresh look. I have pushed some further refactoring to hopefully make it a bit more understandable.